### PR TITLE
feature(amp-img): add amp-img fallback

### PR DESCRIPTION
**Pull request by erwinmombay**
Closes https://github.com/ampproject/amphtml/issues/2335

### CODE REVIEW

**Comment by erwinmombay** [comment]
@dvoytenko PTAL

File: builtins/amp-img.js [hunk 1]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @private {boolean} */
+    this.allowImgLoadFallback_ = false;
+
+    // if this amp-img IS the fallback then don't allow it to have its own
```
Comment (dvoytenko) [inline builtins/amp-img.js]: "if" -> "If"

File: builtins/amp-img.js [hunk 2]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
+    // fallback to stop from nested fallback abuse.
+    if (this.element.hasAttribute('fallback')) {
+      this.allowImgLoadFallback_ = true;
```
Comment (dvoytenko) [inline builtins/amp-img.js]: This looks to do the opposite of what comment says.

File: builtins/amp-img.js [hunk 3]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
-      return true;
-    }
+  /** @override */
+  layoutCallback() {
+    let promise = this.updateImageSrc_();
 
-    /** @override */
-    isRelayoutNeeded() {
-      return true;
+    // we only allow to fallback on error on the initial layoutCallback
```
Comment (dvoytenko) [inline builtins/amp-img.js]: ditto: capitalize comment

File: builtins/amp-img.js [hunk 4]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
-        return Promise.resolve();
-      }
-      const src = this.srcset_.select(this.getLayoutWidth(), this.getDpr()).url;
-      if (src == this.img_.getAttribute('src')) {
-        return Promise.resolve();
-      }
-      this.img_.setAttribute('src', src);
-
-      return loadPromise(this.img_);
+    // clean up the fallback if the src has changed.
```
Comment (dvoytenko) [inline builtins/amp-img.js]: ditto: capitalize all comments please.

File: builtins/amp-img.js [hunk 5]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
+    if (this.allowImgLoadFallback_ &&
```
Comment (dvoytenko) [inline builtins/amp-img.js]: It seems that `allowImgLoadFallback` value is inversed everywhere?

File: builtins/amp-img.js [hunk 6]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
     }
-  };
 
+    this.img_.setAttribute('src', src);
+
+    return loadPromise(this.img_);
+  }
+
+  onImgLoadingError_() {
+    this.getVsync().mutate(() => {
```
Comment (dvoytenko) [inline builtins/amp-img.js]: Should the original error be rethrown/logged?

File: builtins/amp-img.md [hunk 7]
```diff
@@ -36,7 +36,8 @@ limitations under the License.
...
 
 `amp-img` components, like all externally fetched AMP resources, must be given an
-explicit size (as in width / height) in advance, so that the aspect ratio can be known without fetching the image. Actual layout behavior is determined by the layout attribute. If the resource requested by the `amp-img` component fails to load, the space will be blank. A placeholder background color or other visual can be set using CSS selector and style on the element itself.
+explicit size (as in width / height) in advance, so that the aspect ratio can be known without fetching the image. Actual layout behavior is determined by the layout attribute. If the resource requested by the `amp-img` component fails to load, the space will be blank unless a [`fallback`](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md#fallback) child is provided. A fallback is only executed on the initial layout and subsequent src changes after the fact (through resize + srcset for example) will not have a fallback for performance implications.
```
Comment (dvoytenko) [inline builtins/amp-img.md]: "If the resource requested... . fails to load" - this warrants its own paragraph now.

File: builtins/amp-img.js [hunk 8]
```diff
@@ -21,72 +21,103 @@ import {registerElement} from '../src/custom-element';
...
-        return Promise.resolve();
-      }
-      this.img_.setAttribute('src', src);
-
-      return loadPromise(this.img_);
+    // clean up the fallback if the src has changed.
+    if (this.allowImgLoadFallback_ &&
+        this.img_.classList.contains('-amp-ghost')) {
+      this.img_.classList.remove('-amp-ghost');
+      this.toggleFallback(false);
```
Comment (dvoytenko) [inline builtins/amp-img.js]: Setting fallback on the initial layout only makes sense - but we should not reset it immediately before we can confirm that the subsequent layout has succeeded.

File: builtins/amp-img.js (hunk 2 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: woops, i did a search and replace and forgot to change the logic. fixing.

File: builtins/amp-img.js (hunk 5 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: fixed.

File: builtins/amp-img.js (hunk 4 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: done.

File: builtins/amp-img.js (hunk 2 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: done.

File: builtins/amp-img.js (hunk 3 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: done.

File: builtins/amp-img.js (hunk 6 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: good question. imo it should be

File: builtins/amp-img.js (hunk 1 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: done.

File: builtins/amp-img.js (hunk 8 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: @dvoytenko how do we go about checking?

File: builtins/amp-img.md (hunk 7 above)
Comment (erwinmombay) [inline builtins/amp-img.md]: done.

File: builtins/amp-img.js (hunk 6 above)
Comment (erwinmombay) [inline builtins/amp-img.js]: done. rethrown in the `catch`

File: builtins/amp-img.js (hunk 8 above)
Comment (dvoytenko) [inline builtins/amp-img.js]: I think that this piece of code should be done in the `then` response of the layout promise. Similar to how the fallback is set in the `catch` part of the layout promise. The only difference is that `catch` needs to be done only once, while you need to check success and fallback reset for other layouts as well - but internal state can make it an easy check. E.g. you can have this state in the `isFallbackOn_` boolean variable or such to make it faster.

Comment (dvoytenko) [inline builtins/amp-img.js]: Or you can simply keep on checking on the "ghost" class as you do now - I think performance will be acceptable.

Comment (erwinmombay) [inline builtins/amp-img.js]: done. added it on the promise returned from the `updateImgSrc_` funtion, can't add it on the layoutCallback as we can't distinguish if it a promise created from `loadPromise` (we actually do a fetch) or a promise created from `Promise.resolve()` which makes the `updateImgSrc_` function a noop.

**Comment by erwinmombay** [comment]
@dvoytenko PTAL

File: builtins/amp-img.js [hunk 9]
```diff
@@ -21,72 +21,112 @@ import {registerElement} from '../src/custom-element';
...
-      const src = this.srcset_.select(this.getLayoutWidth(), this.getDpr()).url;
-      if (src == this.img_.getAttribute('src')) {
-        return Promise.resolve();
+    this.img_.setAttribute('src', src);
+
+    return this.loadPromise_(this.img_).then(() => {
+      // Clean up the fallback if the src has changed.
+      if (!this.allowImgLoadFallback_ &&
+          this.img_.classList.contains('-amp-ghost')) {
+        this.img_.classList.remove('-amp-ghost');
```
Comment (dvoytenko) [inline builtins/amp-img.js]: `vsync` since we are out of phase here with a micro-task.

Comment (erwinmombay) [inline builtins/amp-img.js]: done.

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**